tile_id_pass/tile_id_pass.py

```diff
@@ -0,0 +1,57 @@
+from gemstone.common.transform import replace, Generator, FromMagma
+from io_core.io_core_magma import IOCore
+from canal.interconnect import Interconnect
+from gemstone.common.configurable import Configurable, ConfigurationType
+import magma
+import mantle
+from gemstone.generator.const import Const
+from hwtypes import BitVector
+
+## We create constant tie hi and tie lo outputs on each tile
+## These pins will be interleaved with the tile_id pins as follows
+## | hi[0] | id[0] | lo[0] | id[1] | hi[1] | id[2] | lo[1] | id[3] | hi[2] | ...
+## We are doing this so that we can actually connect the tile_id inputs
+## to the proper constant values without uniquifying the tiles and while
+## maintaining our abutted array of tile hard macros
+def tile_id_physical(interconnect: Interconnect):
+    tile_id_width = interconnect.tile_id_width
+    tie_hi_width = (tile_id_width // 2) + 1
+    if (tile_id_width % 2) == 0 :
+        tie_lo_width = tile_id_width // 2
+    else:
+        tie_lo_width = (tile_id_width // 2) + 1
+    for (x, y) in interconnect.tile_circuits:
+        tile = interconnect.tile_circuits[(x, y)]
+        tile_core = tile.core
+        if isinstance(tile_core, IOCore) or tile_core is None:
+            continue
+        tile.add_ports(
+            hi=magma.Out(magma.Bits[tie_hi_width]),
+            lo=magma.Out(magma.Bits[tie_lo_width])
+        )
+        # wire all hi bits high
+        tile.wire(tile.ports.hi, Const((2 ** tie_hi_width) - 1))
+        # wire all lo bits low
+        tile.wire(tile.ports.lo, Const(0))
+        # Get the correct tile_id value
+        x_bv = BitVector[tile_id_width / 2](x)
+        y_bv = BitVector[tile_id_width / 2](y)
+        tile_id_bv = BitVector.concat(x_bv, y_bv)
+        # Disconnect existing constant from tile_id port
+        tile_id_port = tile.ports.tile_id
+        for wire in interconnect.wires:
```

Comment:
I really shouldn't have to iterate over all the wires to do this, but all my attempts to remove the wire from the port to the constant value failed. I tried the following:

```
tile_id_port = tile.ports.tile_id
        interconnect.remove_wire(tile_id_port, Const(tile_id_bv))
```
But it didn't work...

Reply:
`Const()` will return a new object, that's why it fails. In the future we need to re-write the gemstone backend to make removing wires easier.